tests/api/test_cluster_descriptor.cpp

```diff
@@ -0,0 +1,140 @@
+    }
+
+    // TODO: Make this test work on a host system without any tt devices.
+    if (pci_device_ids.empty()) {
+        std::cout << "No Tenstorrent devices found. Skipping test." << std::endl;
+        return nullptr;
+    }
+
+    // TODO: remove getting manually cluster descriptor from yaml.
+    std::string yaml_path = test_utils::GetClusterDescYAML();
+    // TODO: Remove the need to do this, allow default constructor to construct with all chips.
+    std::unique_ptr<tt_ClusterDescriptor> cluster_desc;
+    if (device_arch == tt::ARCH::GRAYSKULL) {
+        cluster_desc = tt_ClusterDescriptor::create_for_grayskull_cluster(pci_device_ids_set, pci_device_ids);
+    } else {
+        cluster_desc = tt_ClusterDescriptor::create_from_yaml(yaml_path);
+    }
+
+    return cluster_desc;
+}
+
+TEST(ApiClusterDescriptorTest, BasicFunctionality) {
+
+    std::unique_ptr<tt_ClusterDescriptor> cluster_desc = get_cluster_desc();
+
+    if (cluster_desc == nullptr) {
+        return;
+    }
+
+    std::unordered_set<chip_id_t> all_chips = cluster_desc->get_all_chips();
+    std::unordered_map<chip_id_t, std::uint32_t> harvesting_for_chips = cluster_desc->get_harvesting_info();
+    std::unordered_map<chip_id_t, eth_coord_t> eth_chip_coords = cluster_desc->get_chip_locations();
+    std::unordered_map<chip_id_t, chip_id_t> local_chips_to_pci_device_id = cluster_desc->get_chips_with_mmio();
+    std::unordered_set<chip_id_t> local_chips;
+    for (auto [chip, _]: local_chips_to_pci_device_id) {
+        local_chips.insert(chip);
+    }
+    std::unordered_set<chip_id_t> remote_chips;
+    for (auto chip : all_chips) {
+        if (local_chips.find(chip) == local_chips.end()) {
+            remote_chips.insert(chip);
+        }
+    }
+}
+
+// A standard disjoint set data structure to track connected components.
+class DisjointSet {
+    public:
+        void add_item(int item) {
+            parent[item] = item;
+        }
+
+        int get_parent(int item) {
+            while (parent[item] != item) {
+                item = parent[item];
+            }
+            return item;
+        }
+
+        void merge(int item1, int item2) {
+            int parent1 = get_parent(item1);
+            int parent2 = get_parent(item2);
+            parent[parent1] = parent2;
+        }
+
+        bool are_same_set(int item1, int item2) {
+            return get_parent(item1) == get_parent(item2);
+        }
+
+        int get_num_sets() {
+            std::unordered_set<int> sets;
+            for (auto [item, _]: parent) {
+                sets.insert(get_parent(item));
+            }
+            return sets.size();
+        }
+
+    private:
+        std::unordered_map<int, int> parent;
+};
+
+// This tests fails on a machine with multiple cards.
+// It works as long as all the devices that are discoverable are connected through ethernet.
+// Our ClusterDescriptor doesn't have a notion of multiple unconnected clusters of cards.
```

Comment:
I thought this does work when we have multiple N300s in a system for e.g.

Reply:
This is what I asked about a couple of days ago on the umd_improvement chat if you remember.
Imagine first N300 resolves to chips 0, 1 and the second one resolves to 2, 3.
create-eth-map would give eth coordinates of 0,0,0,0 to 0 and 2, and 1,0,0,0 to 1 and 3.

When calling get_closest_mmio, such as this call in tt_metal: https://github.com/tenstorrent/tt-metal/blob/main/tt_metal/llrt/tt_cluster.cpp#L180 , which fills up devices_grouped_by_assoc_mmio_device_ used later, it will return wrong chip, since it will only look at eth_coords.
For instance, it will return either 0 for both 1 and 3, or 2 for both 1 and 3 (depending on the order of collection).

The tricky thing is that if you have some sequential test going over all remote chips and doing something, it might actually look like it is working, but you'd communicate twice with one of the remote chips actually. And I don't think if we even have a test which would start an AI workload over two N300 cards?

Also note that all of this is not try if those two N300 are connected through an ethernet cable, they would have different eth_coords in that case.